File: vacker_automation/vacker_automation/page/landlord_management_dashboard/landlord_management_dashboard.py
import frappe
from frappe import _
from frappe.utils import today, getdate, add_months, add_days
import json

def get_context(context):
    """Get context for the dashboard"""
    context.no_cache = 1
    context.title = _("Landlord Management Dashboard")
    
    # Get dashboard data
    context.dashboard_data = get_dashboard_data()

def get_dashboard_data():
    """Get comprehensive dashboard data"""
    return {
        "summary": get_summary_data(),
        "charts": get_chart_data(),
        "recent_activities": get_recent_activities(),
        "upcoming_payments": get_upcoming_payments(),
        "contract_expiries": get_contract_expiries(),
        "maintenance_schedules": get_maintenance_schedules(),
        "top_landlords": get_top_landlords()
    }

def get_summary_data():
    """Get summary statistics"""
    try:
        # Total landlords (submitted)
        total_landlords = frappe.db.count("Landlord", {"docstatus": 1})
        frappe.logger().info(f"Total landlords: {total_landlords}")
        
        # Active landlords (have active contracts)
        active_landlords = frappe.db.sql("""
            SELECT COUNT(DISTINCT l.name) 
            FROM `tabLandlord` l
            JOIN `tabLandlord Property` lp ON l.name = lp.parent
            WHERE l.docstatus = 1 
            AND lp.contract_end_date >= %s
            AND lp.status = 'Active'
        """, today())[0][0] or 0
        frappe.logger().info(f"Active landlords: {active_landlords}")
        
        total_properties = frappe.db.count("Property")
        occupied_properties = frappe.db.count("Property", {"property_status": "Occupied"})
        
        # Calculate monthly rental income from active contracts
        monthly_income = frappe.db.sql("""
            SELECT SUM(lp.rental_amount) 
            FROM `tabLandlord` l
            JOIN `tabLandlord Property` lp ON l.name = lp.parent
            WHERE l.docstatus = 1 
            AND lp.contract_end_date >= %s
            AND lp.status = 'Active'
            AND lp.payment_frequency = 'Monthly'
        """, today())[0][0] or 0
        frappe.logger().info(f"Monthly income: {monthly_income}")
        
        quarterly_income = frappe.db.sql("""
            SELECT SUM(lp.rental_amount) 
            FROM `tabLandlord` l
            JOIN `tabLandlord Property` lp ON l.name = lp.parent
            WHERE l.docstatus = 1 
            AND lp.contract_end_date >= %s
            AND lp.status = 'Active'
            AND lp.payment_frequency = 'Quarterly'
        """, today())[0][0] or 0
        frappe.logger().info(f"Quarterly income: {quarterly_income}")
        
        annual_income = frappe.db.sql("""
            SELECT SUM(lp.rental_amount) 
            FROM `tabLandlord` l
            JOIN `tabLandlord Property` lp ON l.name = lp.parent
            WHERE l.docstatus = 1 
            AND lp.contract_end_date >= %s
            AND lp.status = 'Active'
            AND lp.payment_frequency = 'Annually'
        """, today())[0][0] or 0
        frappe.logger().info(f"Annual income: {annual_income}")
        
        total_monthly_income = monthly_income + (quarterly_income / 3) + (annual_income / 12)
        frappe.logger().info(f"Total monthly income: {total_monthly_income}")
        
        # Get overdue payments
        overdue_payments = frappe.db.count("Landlord Payment Schedule", {
            "status": "Overdue"
        })
        frappe.logger().info(f"Overdue payments: {overdue_payments}")
        
        # Get contracts expiring this month
        contracts_expiring = frappe.db.sql("""
            SELECT COUNT(*) 
            FROM `tabLandlord` l
            JOIN `tabLandlord Property` lp ON l.name = lp.parent
            WHERE l.docstatus = 1
            AND lp.contract_end_date BETWEEN %s AND %s
            AND lp.status = 'Active'
        """, [today(), add_months(today(), 1)])[0][0] or 0
        frappe.logger().info(f"Contracts expiring: {contracts_expiring}")
        
        summary_data = {
            "total_landlords": total_landlords,
            "active_landlords": active_landlords,
            "total_properties": total_properties,
            "occupied_properties": occupied_properties,
            "monthly_income": total_monthly_income,
            "overdue_payments": overdue_payments,
            "contracts_expiring": contracts_expiring
        }
        
        frappe.logger().info(f"Summary data: {summary_data}")
        return summary_data
        
    except Exception as e:
        frappe.logger().error(f"Error in get_summary_data: {str(e)}")
        raise

def get_chart_data():
    """Get chart data for dashboard"""
    try:
        # Revenue by landlord type
        revenue_by_type = frappe.db.sql("""
            SELECT 
                l.landlord_type,
                SUM(CASE 
                    WHEN lp.payment_frequency = 'Monthly' THEN lp.rental_amount * 12
                    WHEN lp.payment_frequency = 'Quarterly' THEN lp.rental_amount * 4
                    WHEN lp.payment_frequency = 'Annually' THEN lp.rental_amount
                    ELSE 0
                END) as annual_revenue
            FROM `tabLandlord` l
            JOIN `tabLandlord Property` lp ON l.name = lp.parent
            WHERE l.docstatus = 1 
            AND lp.contract_end_date >= %s
            AND lp.status = 'Active'
            GROUP BY l.landlord_type
        """, today(), as_dict=1)
        frappe.logger().info(f"Revenue by type: {revenue_by_type}")
        
        # Revenue by media type
        revenue_by_media = frappe.db.sql("""
            SELECT 
                lp.media_type,
                SUM(CASE 
                    WHEN lp.payment_frequency = 'Monthly' THEN lp.rental_amount * 12
                    WHEN lp.payment_frequency = 'Quarterly' THEN lp.rental_amount * 4
                    WHEN lp.payment_frequency = 'Annually' THEN lp.rental_amount
                    ELSE 0
                END) as annual_revenue
            FROM `tabLandlord` l
            JOIN `tabLandlord Property` lp ON l.name = lp.parent
            WHERE l.docstatus = 1 
            AND lp.contract_end_date >= %s
            AND lp.status = 'Active'
            GROUP BY lp.media_type
        """, today(), as_dict=1)
        frappe.logger().info(f"Revenue by media: {revenue_by_media}")
        
        # Property status distribution
        property_status = frappe.db.sql("""
            SELECT property_status, COUNT(*) as count
            FROM `tabProperty`
            GROUP BY property_status
        """, as_dict=1)
        frappe.logger().info(f"Property status: {property_status}")
        
        chart_data = {
            "revenue_by_type": revenue_by_type,
            "revenue_by_media": revenue_by_media,
            "property_status": property_status
        }
        
        frappe.logger().info(f"Chart data: {chart_data}")
        return chart_data
        
    except Exception as e:
        frappe.logger().error(f"Error in get_chart_data: {str(e)}")
        raise

def get_recent_activities():
    """Get recent activities"""
    recent_landlords = frappe.db.sql("""
        SELECT name, full_legal_name, date_of_onboarding, landlord_type
        FROM `tabLandlord`
        WHERE docstatus = 1
        ORDER BY creation DESC
        LIMIT 5
    """, as_dict=1)
    
    recent_payments = frappe.db.sql("""
        SELECT lps.name, lps.landlord, lps.property, lps.amount, lps.payment_date
        FROM `tabLandlord Payment Schedule` lps
        WHERE lps.status = 'Paid'
        ORDER BY lps.payment_date DESC
        LIMIT 5
    """, as_dict=1)
    
    return {
        "recent_landlords": recent_landlords,
        "recent_payments": recent_payments
    }

def get_upcoming_payments():
    """Get upcoming payments"""
    upcoming_payments = frappe.db.sql("""
        SELECT 
            lps.name,
            lps.landlord,
            lps.property,
            lps.amount,
            lps.due_date,
            l.full_legal_name
        FROM `tabLandlord Payment Schedule` lps
        JOIN `tabLandlord` l ON lps.landlord = l.name
        WHERE lps.status = 'Pending'
        AND lps.due_date BETWEEN %s AND %s
        ORDER BY lps.due_date ASC
        LIMIT 10
    """, [today(), add_months(today(), 1)], as_dict=1)
    
    return upcoming_payments

def get_contract_expiries():
    """Get contracts expiring soon"""
    expiring_contracts = frappe.db.sql("""
        SELECT 
            l.name,
            l.full_legal_name,
            lp.property,
            lp.contract_end_date,
            lp.rental_amount,
            lp.media_type
        FROM `tabLandlord` l
        JOIN `tabLandlord Property` lp ON l.name = lp.parent
        WHERE l.docstatus = 1
        AND lp.contract_end_date BETWEEN %s AND %s
        AND lp.status = 'Active'
        ORDER BY lp.contract_end_date ASC
        LIMIT 10
    """, [today(), add_months(today(), 3)], as_dict=1)
    
    return expiring_contracts

def get_maintenance_schedules():
    """Get upcoming maintenance schedules"""
    maintenance_schedules = frappe.db.sql("""
        SELECT 
            ms.name,
            ms.landlord,
            ms.property,
            ms.scheduled_date,
            ms.maintenance_type,
            l.full_legal_name
        FROM `tabMaintenance Schedule` ms
        JOIN `tabLandlord` l ON ms.landlord = l.name
        WHERE ms.status = 'Scheduled'
        AND ms.scheduled_date BETWEEN %s AND %s
        ORDER BY ms.scheduled_date ASC
        LIMIT 10
    """, [today(), add_months(today(), 1)], as_dict=1)
    
    return maintenance_schedules

def get_top_landlords():
    """Get top landlords by revenue"""
    top_landlords = frappe.db.sql("""
        SELECT 
            l.name,
            l.full_legal_name,
            SUM(CASE 
                WHEN lp.payment_frequency = 'Monthly' THEN lp.rental_amount * 12
                WHEN lp.payment_frequency = 'Quarterly' THEN lp.rental_amount * 4
                WHEN lp.payment_frequency = 'Annually' THEN lp.rental_amount
                ELSE 0
            END) as annual_revenue,
            COUNT(lp.property) as property_count
        FROM `tabLandlord` l
        JOIN `tabLandlord Property` lp ON l.name = lp.parent
        WHERE l.docstatus = 1 
        AND lp.contract_end_date >= %s
        AND lp.status = 'Active'
        GROUP BY l.name, l.full_legal_name
        ORDER BY annual_revenue DESC
        LIMIT 10
    """, today(), as_dict=1)
    
    return top_landlords

@frappe.whitelist()
def get_dashboard_stats():
    """API endpoint for dashboard statistics"""
    try:
        dashboard_data = get_dashboard_data()
        frappe.logger().info(f"Dashboard data: {dashboard_data}")
        return dashboard_data
    except Exception as e:
        frappe.logger().error(f"Error getting dashboard stats: {str(e)}")
        frappe.throw(f"Error getting dashboard stats: {str(e)}")

@frappe.whitelist()
def get_landlord_details(landlord_name):
    """Get detailed landlord information"""
    landlord = frappe.get_doc("Landlord", landlord_name)
    return landlord.get_landlord_summary()

@frappe.whitelist()
def get_property_details(property_name):
    """Get detailed property information"""
    property_doc = frappe.get_doc("Property", property_name)
    return property_doc.get_property_details() 

File: vacker_automation/vacker_automation/page/landlord_management_dashboard/landlord_management_dashboard.js
// Landlord Management Dashboard JavaScript
frappe.pages['landlord-management-dashboard'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: __('Landlord Management Dashboard'),
        single_column: true
    });

    // Initialize dashboard
    init_dashboard(page);
};

function init_dashboard(page) {
    // Create dashboard layout
    const dashboard = $(`
        <div class="landlord-dashboard">
            <div class="dashboard-header">
                <h1>Landlord Management Dashboard</h1>
                <div class="dashboard-actions">
                    <button class="btn btn-primary" onclick="refreshDashboard()">
                        <i class="fa fa-refresh"></i> Refresh
                    </button>
                    <button class="btn btn-success" onclick="exportDashboard()">
                        <i class="fa fa-download"></i> Export
                    </button>
                </div>
            </div>
            
            <div class="dashboard-summary">
                <div class="summary-card" id="total-landlords">
                    <div class="card-icon">
                        <i class="fa fa-users"></i>
                    </div>
                    <div class="card-content">
                        <h3>Total Landlords</h3>
                        <div class="card-value" id="total-landlords-value">-</div>
                    </div>
                </div>
                
                <div class="summary-card" id="active-landlords">
                    <div class="card-icon">
                        <i class="fa fa-user-check"></i>
                    </div>
                    <div class="card-content">
                        <h3>Active Landlords</h3>
                        <div class="card-value" id="active-landlords-value">-</div>
                    </div>
                </div>
                
                <div class="summary-card" id="monthly-income">
                    <div class="card-icon">
                        <i class="fa fa-money"></i>
                    </div>
                    <div class="card-content">
                        <h3>Monthly Income</h3>
                        <div class="card-value" id="monthly-income-value">-</div>
                    </div>
                </div>
                
                <div class="summary-card" id="overdue-payments">
                    <div class="card-icon">
                        <i class="fa fa-exclamation-triangle"></i>
                    </div>
                    <div class="card-content">
                        <h3>Overdue Payments</h3>
                        <div class="card-value" id="overdue-payments-value">-</div>
                    </div>
                </div>
            </div>
            
            <div class="dashboard-charts">
                <div class="chart-container">
                    <h3>Revenue by Landlord Type</h3>
                    <div id="revenue-by-type-chart"></div>
                </div>
                
                <div class="chart-container">
                    <h3>Revenue by Media Type</h3>
                    <div id="revenue-by-media-chart"></div>
                </div>
                
                <div class="chart-container">
                    <h3>Property Status Distribution</h3>
                    <div id="property-status-chart"></div>
                </div>
            </div>
            
            <div class="dashboard-details">
                <div class="detail-section">
                    <h3>Upcoming Payments</h3>
                    <div id="upcoming-payments-list"></div>
                </div>
                
                <div class="detail-section">
                    <h3>Contracts Expiring Soon</h3>
                    <div id="contract-expiries-list"></div>
                </div>
                
                <div class="detail-section">
                    <h3>Top Landlords by Revenue</h3>
                    <div id="top-landlords-list"></div>
                </div>
            </div>
        </div>
    `);

    page.main.append(dashboard);
    
    // Load dashboard data
    loadDashboardData();
    
    // Set up auto-refresh
    setInterval(loadDashboardData, 300000); // Refresh every 5 minutes
}

function loadDashboardData() {
    frappe.call({
        method: 'vacker_automation.vacker_automation.page.landlord_management_dashboard.landlord_management_dashboard.get_dashboard_stats',
        callback: function(r) {
            console.log('Dashboard data received:', r);
            if (r.message) {
                updateDashboard(r.message);
            } else {
                console.error('No dashboard data received');
                showError('Failed to load dashboard data');
            }
        },
        error: function(err) {
            console.error('Error loading dashboard data:', err);
            showError('Error loading dashboard data: ' + err.message);
        }
    });
}

function updateDashboard(data) {
    console.log('Updating dashboard with data:', data);
    
    // Update summary cards
    if (data.summary) {
        updateSummaryCards(data.summary);
    } else {
        console.warn('No summary data available');
    }
    
    // Update charts
    if (data.charts) {
        updateCharts(data.charts);
    } else {
        console.warn('No chart data available');
    }
    
    // Update detail sections
    updateDetailSections(data);
}

function updateSummaryCards(summary) {
    console.log('Updating summary cards with:', summary);
    
    $('#total-landlords-value').text(summary.total_landlords || 0);
    $('#active-landlords-value').text(summary.active_landlords || 0);
    $('#monthly-income-value').text(formatCurrency(summary.monthly_income || 0));
    $('#overdue-payments-value').text(summary.overdue_payments || 0);
}

function formatCurrency(amount) {
    if (amount === null || amount === undefined || isNaN(amount)) return '0.00';
    return parseFloat(amount).toFixed(2);
}

function updateCharts(charts) {
    console.log('Updating charts with:', charts);
    
    // Revenue by landlord type chart
    if (charts.revenue_by_type && charts.revenue_by_type.length > 0) {
        const chartData = charts.revenue_by_type.map(item => ({
            name: item.landlord_type || 'Unknown',
            value: parseFloat(item.annual_revenue || 0)
        }));
        
        renderPieChart('#revenue-by-type-chart', chartData, 'Revenue by Landlord Type');
    } else {
        console.warn('No revenue by type data available');
        $('#revenue-by-type-chart').html('<div class="no-data">No revenue data available</div>');
    }
    
    // Revenue by media type chart
    if (charts.revenue_by_media && charts.revenue_by_media.length > 0) {
        const chartData = charts.revenue_by_media.map(item => ({
            name: item.media_type || 'Unknown',
            value: parseFloat(item.annual_revenue || 0)
        }));
        
        renderPieChart('#revenue-by-media-chart', chartData, 'Revenue by Media Type');
    } else {
        console.warn('No revenue by media data available');
        $('#revenue-by-media-chart').html('<div class="no-data">No revenue data available</div>');
    }
    
    // Property status chart
    if (charts.property_status && charts.property_status.length > 0) {
        const chartData = charts.property_status.map(item => ({
            name: item.property_status || 'Unknown',
            value: parseInt(item.count || 0)
        }));
        
        renderPieChart('#property-status-chart', chartData, 'Property Status Distribution');
    } else {
        console.warn('No property status data available');
        $('#property-status-chart').html('<div class="no-data">No property data available</div>');
    }
}

function renderPieChart(container, data, title) {
    const chartContainer = $(container);
    chartContainer.empty();
    
    if (data.length === 0) {
        chartContainer.html('<div class="no-data">No data available</div>');
        return;
    }
    
    // Simple pie chart using HTML/CSS
    const total = data.reduce((sum, item) => sum + item.value, 0);
    let currentAngle = 0;
    
    const chart = $('<div class="pie-chart"></div>');
    
    data.forEach((item, index) => {
        const percentage = total > 0 ? (item.value / total) * 100 : 0;
        const angle = (percentage / 100) * 360;
        
        const slice = $(`
            <div class="pie-slice" style="
                transform: rotate(${currentAngle}deg);
                background: ${getChartColor(index)};
                clip-path: polygon(50% 50%, 50% 0%, ${50 + 50 * Math.cos(angle * Math.PI / 180)}% ${50 - 50 * Math.sin(angle * Math.PI / 180)}%);
            "></div>
        `);
        
        chart.append(slice);
        currentAngle += angle;
    });
    
    chartContainer.append(chart);
    
    // Add legend
    const legend = $('<div class="chart-legend"></div>');
    data.forEach((item, index) => {
        const legendItem = $(`
            <div class="legend-item">
                <span class="legend-color" style="background: ${getChartColor(index)}"></span>
                <span class="legend-label">${item.name}</span>
                <span class="legend-value">${formatCurrency(item.value)}</span>
            </div>
        `);
        legend.append(legendItem);
    });
    
    chartContainer.append(legend);
}

function getChartColor(index) {
    const colors = [
        '#FF6B6B', '#4ECDC4', '#45B7D1', '#96CEB4', '#FFEAA7',
        '#DDA0DD', '#98D8C8', '#F7DC6F', '#BB8FCE', '#85C1E9'
    ];
    return colors[index % colors.length];
}

function updateDetailSections(data) {
    // Update upcoming payments
    updateUpcomingPayments(data.upcoming_payments);
    
    // Update contract expiries
    updateContractExpiries(data.contract_expiries);
    
    // Update top landlords
    updateTopLandlords(data.top_landlords);
}

function updateUpcomingPayments(payments) {
    const container = $('#upcoming-payments-list');
    container.empty();
    
    if (!payments || payments.length === 0) {
        container.html('<div class="no-data">No upcoming payments</div>');
        return;
    }
    
    const table = $('<table class="table table-striped"></table>');
    table.append(`
        <thead>
            <tr>
                <th>Landlord</th>
                <th>Property</th>
                <th>Amount</th>
                <th>Due Date</th>
            </tr>
        </thead>
    `);
    
    const tbody = $('<tbody></tbody>');
    payments.forEach(payment => {
        const row = $(`
            <tr>
                <td>${payment.full_legal_name || payment.landlord}</td>
                <td>${payment.property}</td>
                <td>${formatCurrency(payment.amount)}</td>
                <td>${frappe.format(payment.due_date, {fieldtype: 'Date'})}</td>
            </tr>
        `);
        tbody.append(row);
    });
    
    table.append(tbody);
    container.append(table);
}

function updateContractExpiries(contracts) {
    const container = $('#contract-expiries-list');
    container.empty();
    
    if (!contracts || contracts.length === 0) {
        container.html('<div class="no-data">No contracts expiring soon</div>');
        return;
    }
    
    const table = $('<table class="table table-striped"></table>');
    table.append(`
        <thead>
            <tr>
                <th>Landlord</th>
                <th>Property</th>
                <th>End Date</th>
                <th>Rental Amount</th>
            </tr>
        </thead>
    `);
    
    const tbody = $('<tbody></tbody>');
    contracts.forEach(contract => {
        const row = $(`
            <tr>
                <td>${contract.full_legal_name}</td>
                <td>${contract.property}</td>
                <td>${frappe.format(contract.contract_end_date, {fieldtype: 'Date'})}</td>
                <td>${formatCurrency(contract.rental_amount)}</td>
            </tr>
        `);
        tbody.append(row);
    });
    
    table.append(tbody);
    container.append(table);
}

function updateTopLandlords(landlords) {
    const container = $('#top-landlords-list');
    container.empty();
    
    if (!landlords || landlords.length === 0) {
        container.html('<div class="no-data">No landlord data available</div>');
        return;
    }
    
    const table = $('<table class="table table-striped"></table>');
    table.append(`
        <thead>
            <tr>
                <th>Landlord</th>
                <th>Property</th>
                <th>Annual Revenue</th>
                <th>Payment Frequency</th>
            </tr>
        </thead>
    `);
    
    const tbody = $('<tbody></tbody>');
    landlords.forEach(landlord => {
        const row = $(`
            <tr>
                <td>${landlord.full_legal_name}</td>
                <td>${landlord.property}</td>
                <td>${formatCurrency(landlord.annual_revenue)}</td>
                <td>${landlord.payment_frequency}</td>
            </tr>
        `);
        tbody.append(row);
    });
    
    table.append(tbody);
    container.append(table);
}

function refreshDashboard() {
    loadDashboardData();
    frappe.show_alert('Dashboard refreshed successfully', 3);
}

function exportDashboard() {
    frappe.call({
        method: 'vacker_automation.vacker_automation.page.landlord_management_dashboard.landlord_management_dashboard.get_dashboard_stats',
        callback: function(r) {
            if (r.message) {
                // Create CSV export
                const csv = convertToCSV(r.message);
                downloadCSV(csv, 'landlord-dashboard-export.csv');
            }
        }
    });
}

function convertToCSV(data) {
    // Implementation for CSV conversion
    // This is a simplified version - you might want to use a proper CSV library
    let csv = 'Landlord Management Dashboard Export\n\n';
    
    // Summary data
    csv += 'Summary Statistics\n';
    csv += 'Total Landlords,' + data.summary.total_landlords + '\n';
    csv += 'Active Landlords,' + data.summary.active_landlords + '\n';
    csv += 'Monthly Income,' + data.summary.monthly_income + '\n';
    csv += 'Overdue Payments,' + data.summary.overdue_payments + '\n\n';
    
    return csv;
}

function downloadCSV(csv, filename) {
    const blob = new Blob([csv], { type: 'text/csv' });
    const url = window.URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = filename;
    a.click();
    window.URL.revokeObjectURL(url);
}

function showError(message) {
    // Show error message on dashboard
    const errorDiv = $('<div class="alert alert-danger">' + message + '</div>');
    $('.landlord-dashboard').prepend(errorDiv);
    
    // Remove error after 5 seconds
    setTimeout(() => {
        errorDiv.fadeOut();
    }, 5000);
} 